Add post announcing pulumi-kubernetes-operator (#3902)
pulumi/docs: content/blog/pulumi-kubernetes-operator/index.md

---
date: "2020-08-12"
title: "Introducing the Pulumi Kubernetes Operator"
authors: ["mike-metral"]
tags: ["Kubernetes", "Operator", "Continuous Delivery"]
meta_desc: "Introducing the Pulumi Kubernetes Operator: Deploy infrastructure in Pulumi Stacks"
meta_image: operator.png
---

Kubernetes developers and operators work together to manage workloads

> and to continuously ship software through CI/CD. These users

Reviewer: ```suggestion
and continuously ship software through CI/CD. Users
```

> have an affinity for automation and pipelines, and richer integration

Reviewer: ```suggestion
with an affinity for automation and pipelines and richer integration
```

> with Kubernetes is a growing theme across the cloud native ecosystem.

Reviewer: ```suggestion
with Kubernetes is a growing theme across the cloud-native ecosystem.
```

We're excited to introduce the Pulumi Kubernetes Operator: a Kubernetes
controller that deploys cloud infrastructure in Pulumi Stacks for you and your
team.

These program stacks include virtual machines, block storage,

> managed Kuberentes clusters, API resources, serverless functions and more!

Reviewer: ```suggestion
managed Kubernetes clusters, API resources, serverless functions, and more!
```

<!--more-->

## Overview

The [Pulumi Kubernetes Operator][pulumi-k8s-op] is an [extension pattern][k8s-ext-pattern] that

> enables Kuberentes users to create a Pulumi [`Stack`][stack] as a first-class API

Reviewer: ```suggestion
enables Kubernetes users to create a Pulumi [`Stack`][stack] as a first-class API
```

resource, and use the `StackController` to drive the updates of the Stack until
success.

Check out [how to deploy the operator][deploy-op] to a Kubernetes cluster using any of Pulumi's supported

> languages, or through YAML manifests and `kubectl`.

Reviewer: ```suggestion
languages or YAML manifests and `kubectl`.
```

            }
        });

        // Create an Kubernetes cluster on AWS EKS.
        var myStack = new Pulumi.Kubernetes.ApiExtensions.CustomResource("my-stack", new StackArgs
        {
            Spec = new StackSpecArgs
            {
                Stack = "<YOUR_ORG>/pulumi-aws-eks/dev",
                ProjectRepo = "https://github.com/metral/pulumi-aws-eks",
                Commit = "fc4ab1a3e49d48cf5c764cd8cd626879a90bcc45",
                AccessTokenSecret = accessToken.Metadata.Apply(m => m.Name),
                Config =
                {
                    {"aws:region", "us-west-2"}
                },
                EnvSecrets = {awsCreds.Metadata.Apply(m => m.Name)},
                InitOnCreate = true,
                DestroyOnFinalize = true,
            }
        });
    }
}
```

{{% /choosable %}}

{{% choosable language go %}}

```go
Coming Soon
```

{{% /choosable %}}
{{% /chooser %}}

## CI/CD Scenarios

As you build out your Kubernetes workloads, you'll need to implement update strategies
to ship new software easily and reliably. Kubernetes API resources like
[Deployments][k8s-deployment] are used to model stateless apps and operate their lifecycles

> e.g. recreate app replicas, or perform a rolling update.

Reviewer: ```suggestion
e.g., recreate app replicas, or perform a rolling update.
```

A common approach that most teams rely on is the [Blue / Green][blue-green] deployment.

> This starts with an initial version of the app, and then deploys an updated version

Reviewer: ```suggestion
This starts with an initial version of the app and then deploys an updated version
```

alongside the initial version. Once ready, traffic is switched over from the
initial version to the new version with no downtime.

![Blue Green Kubernetes Deployment](./blue-green.png)

> In Kubernetes, Blue/Green is modeled by using a [Service][k8s-service]

Reviewer: ```suggestion
In Kubernetes, Blue/Green is modeled using a [Service][k8s-service]
```

that load balances and selects a set of Deployment pods labeled "blue," and then switching to a
different selection of updated, "green" pods.

> Similar to the Stack CustomResources in the previous sections, we can deploy a Stack

Reviewer: ```suggestion
Like the Stack CustomResources in the previous sections, we can deploy a Stack
```

of a [Blue/Green Kubernetes app in Pulumi][blue-green-example], and step through a sequence of its Git commits

> similiar to how a CI/CD pipeline does. Additionally, we can use the [Pulumi config system][pulumi-config] to parameterize

Reviewer: ```suggestion
similar to how a CI/CD pipeline does. Additionally, we can use the [Pulumi config system][pulumi-config] to parameterize
```

> settings and secrets on how to manage the Stack CustomResource project.

Reviewer: ```suggestion
settings and secrets to manage the Stack CustomResource project.
```

Deploying the active blue and passive green versions of the app can be done on
the example Stack with the following steps.

```bash
pulumi config set --secret pulumiAccessToken <YOUR_PULUMI_API_TOKEN>
pulumi config set stackProjectRepo https://github.com/metral/pulumi-blue-green
pulumi config set stackCommit b19759220f25476605620fdfffeface39a630246
pulumi up -y
```

After deployment, we'll test that the v1 / "blue" endpoint is live and running.
The v2 / "green" app will be on standby.

```bash
while true; do curl http://34.83.25.150:80/version ; echo; sleep 1; done

{"id":1,"content":"current"}
{"id":1,"content":"current"}
{"id":1,"content":"current"}
{"id":1,"content":"current"}
...
```

When we're ready to switch traffic, the blue/green
transition is done by updating the config to the Git new commit of the selector

> swap, and running an update.

Reviewer: ```suggestion
swap and running an update.
```

```bash
pulumi config set stackCommit f4bf2f7b54ec441d5af374933cca4ef09f2bce24
pulumi up -y
```

> After a few seconds you should start seeing the generated traffic returning the

Reviewer: ```suggestion
After a few seconds, you should start seeing the generated traffic returning the
```

new version in a clean cut-over with no dropped packets.

```bash
...
{"id":1,"content":"current"}
{"id":1,"content":"current"}
{"id":1,"content":"current"}
{"id":2,"content":"new"}
{"id":2,"content":"new"}
{"id":2,"content":"new"}
{"id":2,"content":"new"}
...
```

> Check out the [full walkthrough][blue-green-walkthrough] for a tutorial, and

Reviewer: ```suggestion
Check out the [full walkthrough][blue-green-walkthrough] for a tutorial and
```

the video clip below for a demo.

> TODO: add video-clip blue/green stack demo.

Reviewer 2: Is there a link for this now?

Author: The video is in [Gdrive](https://drive.google.com/drive/u/1/folders/156HJjYlQudFOdrPRrVHokJhPlZnoiRFU). Will be uploaded to YT soon

[blue-green-example]: https://github.com/metral/pulumi-blue-green/blob/master/index.ts
[k8s-deployment]: https://kubernetes.io/docs/concepts/workloads/controllers/deployment
[blue-green]: https://en.wikipedia.org/wiki/Blue-green_deployment
[k8s-service]: https://kubernetes.io/docs/concepts/services-networking/service
[blue-green-walkthrough]: https://github.com/pulumi/pulumi-kubernetes-operator/tree/master/examples/blue-green

## Wrap-Up

The [Pulumi Kubernetes Operator][pulumi-k8s-op] helps deploy Pulumi Stacks in

> your Kubernetes cluster, and naturally fits in with your CI/CD process.

Reviewer: ```suggestion
your Kubernetes cluster and naturally fits your CI/CD process.
```

We covered examples like deploying an [NGINX Stack][pulumi-k8s-nginx] in the same cluster as the
Operator, creating a new Kubernetes cluster in an [AWS EKS Stack][pulumi-aws-eks], and how Stacks can be
abstracted for deployment strategies like Blue/Green.

There are [many more][p-examples] examples of deploying infrastructure,

> containers, serverless, and Kubernetes across cloud providers and cloud native services.

Reviewer: ```suggestion
containers, serverless, and Kubernetes across cloud providers and cloud-native services.
```
